Mathlib/Tactic/Linter/SupersededTacticLinter.lean

```diff
@@ -0,0 +1,85 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa, Jeremy Tan
+-/
+
+import Lean.Elab.Command
+-- Import this linter explicitly to ensure that
+-- this file has a valid copyright header and module docstring.
+import Mathlib.Tactic.Linter.Header
+
+/-!
+# Linter against superseded tactics (refine' and cases')
+
+`refine'` and `cases'` provide backward-compatible implementations of their unprimed equivalents
+in Lean 3, `refine` and `cases`. This linter flags usages of these two tactics, because they have
+been superseded by Lean 4 tactics:
+
+* `refine` replaces `refine'`. While they are similar, they handle metavariables slightly
+  differently; this means that they are not completely interchangeable, nor can one completely
+  replace the other. However, `refine` is more readable and (heuristically) tends to be more
+  efficient on average.
+* `obtain`, `rcases` and `cases` replace `cases'`. Unlike the replacement tactics, `cases'`
+  does not require the variables it introduces to be separated by case, which hinders readability.
+
+This linter is an incentive to discourage uses of such superseded tactics, without being a ban.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter.Style
+
+/-- The "refine" linter flags usages of the `refine'` tactic.
+
+The tactics `refine` and `refine'` are similar, but they handle meta-variables slightly differently.
+This means that they are not completely interchangeable, nor can one completely replace the other.
+However, `refine` is more readable and (heuristically) tends to be more efficient on average.
+-/
+register_option linter.style.refine : Bool := {
+  defValue := false
+  descr := "enable the refine linter"
+}
+
+/-- The "cases" linter flags uses of the `cases'` tactic, which is a backward-compatible version of
+Lean 3's `cases` tactic. Unlike `obtain`, `rcases` and Lean 4's `cases`, variables introduced by
+`cases'` are not required to be separated by case, which hinders readability.
+-/
+register_option linter.style.cases : Bool := {
+  defValue := false
+  descr := "enable the cases linter"
+}
+
+/-- `getSupersededTactics t` returns all usages of superseded tactics in the input syntax `t`. -/
+partial
+def getSupersededTactics : Syntax → Array (Nat × Syntax × MessageData)
```

Comment:
```suggestion
def getSupersededTactics : Syntax → Array (SyntaxNodeKind × Syntax × MessageData)
```
How about using the `SyntaxNodeKind` itself to index what tactic it is?

Reply:
@adomani all done.